`TestProxyMakeConnectionHandler` flakiness

## Failure

#### Link(s) to logs
https://console.cloud.google.com/cloud-build/builds/ed142bd5-a939-4f67-9950-7e9e979d7975?project=ci-account

#### Relevant snippet
```
--- PASS: TestProxyMakeConnectionHandler (9.96s)
panic: Fail in goroutine after TestProxyMakeConnectionHandler has completed

goroutine 449 [running]:
testing.(*common).Fail(0xc000e569c0)
	/opt/go/src/testing/testing.go:817 +0x19b
testing.(*common).Errorf(0xc000e569c0, {0x455678a, 0x3}, {0xc0013863d0, 0x1, 0x1})
	/opt/go/src/testing/testing.go:934 +0xa7
github.com/stretchr/testify/assert.Fail({0x7fc2ab2d85f8, 0xc000e569c0}, {0xc000db4720, 0x2b}, {0x0, 0x0, 0x0})
	/go/pkg/mod/github.com/stretchr/testify@v1.7.1/assert/assertions.go:262 +0x429
github.com/stretchr/testify/assert.NoError({0x7fc2ab2d85f8, 0xc000e569c0}, {0x568c4c0, 0xc0000de2c0}, {0x0, 0x0, 0x0})
	/go/pkg/mod/github.com/stretchr/testify@v1.7.1/assert/assertions.go:1334 +0x125
github.com/stretchr/testify/require.NoError({0x56a8098, 0xc000e569c0}, {0x568c4c0, 0xc0000de2c0}, {0x0, 0x0, 0x0})
	/go/pkg/mod/github.com/stretchr/testify@v1.7.1/require/require.go:1261 +0xa5
github.com/gravitational/teleport/lib/srv/alpnproxy.TestProxyMakeConnectionHandler.func1()
	/workspace/lib/srv/alpnproxy/proxy_test.go:381 +0xa8
created by github.com/gravitational/teleport/lib/srv/alpnproxy.TestProxyMakeConnectionHandler
	/workspace/lib/srv/alpnproxy/proxy_test.go:379 +0x81d
FAIL	github.com/gravitational/teleport/lib/srv/alpnproxy	29.005s
```

> https://console.cloud.google.com/cloud-build/builds/e9e8d732-2679-4acb-af42-ca2b6c68cf9f?project=ci-account
> 
> Another failure with what looks like a different problem from the OP
> 
> ```
> --- FAIL: TestProxyMakeConnectionHandler (10.12s)
>         	Test:       	TestProxyMakeConnectionHandler
>         	            	context canceled
>         	Error:      	Received unexpected error:
>         	            				asm_amd64.s:1571
>         	Error Trace:	proxy_test.go:381
>     proxy_test.go:381: 
> ```

Thanks! There is certainly a race. Will fix now.

--- a/lib/srv/alpnproxy/proxy_test.go
+++ b/lib/srv/alpnproxy/proxy_test.go
@@ -372,13 +372,11 @@ func TestProxyMakeConnectionHandler(t *testing.T) {
 	defer clientConn.Close()
 	defer serverConn.Close()
 
-	timeoutCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
-	t.Cleanup(cancel)
-
 	// Let alpnConnHandler serve the connection in a separate go routine.
+	handlerCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	go func() {
-		alpnConnHandler(timeoutCtx, serverConn)
-		require.NoError(t, timeoutCtx.Err())
+		defer cancel()
+		alpnConnHandler(handlerCtx, serverConn)
 	}()
 
 	// Send client request.
@@ -403,6 +401,10 @@ func TestProxyMakeConnectionHandler(t *testing.T) {
 	response, err := http.ReadResponse(bufio.NewReader(clientTLSConn), req)
 	require.NoError(t, err)
 	require.Equal(t, http.StatusOK, response.StatusCode)
+
+	// Wait until handler is done. And verify context is canceled, NOT deadline exceeded.
+	<-handlerCtx.Done()
+	require.ErrorIs(t, handlerCtx.Err(), context.Canceled)
 }
 
 // TestProxyALPNProtocolsRouting tests the routing based on client TLS NextProtos values.

